PS-7702 Document jemalloc memory allocation profiling on PS 8.0 (8.0) (#4352)
percona/percona-server: doc/source/performance/jemalloc-profiling.rst

.. _jemalloc-profiling:

=================================================
Jemalloc Memory Allocation Profiling
=================================================

> Implemented in :rn:`8.0.25-15`, you are able to profile the jemalloc memory allocation.

Reviewer: I think we need to be clearer and indicate that this is only present and useful if the server is running with the jemalloc library.  @ldonoso ?

Author: Fixed. Completely rewritten.

Reviewer 2: Sorry for the delay @georgelorchpercona, yes, I agree, thanks for the change @patrickbirch

.. _jemalloc-profiling:

=================================================
Jemalloc Memory Allocation Profiling
=================================================

> Implemented in :rn:`8.0.25-15`, |Percona Server| can take advantage of the  memory-profiling ability of the jemalloc allocator. This ability provides a method to investigate memory-related issues.

Reviewer 2: Please, mind the two spaces on
```suggestion
Implemented in :rn:`8.0.25-15`, |Percona Server| can take advantage of the memory-profiling ability of the jemalloc allocator. This ability provides a method to investigate memory-related issues.
```

Author: Fixed.

Requirements
----------------

> This memory-profiling requires :ref:`jemalloc_detected`. This read-only variable returns ``true`` if jemalloc with the profiling-enabled option is detected. 

Reviewer 2: ```suggestion
This memory-profiling requires :ref:`jemalloc_detected`. This read-only variable returns ``true`` if jemalloc with the profiling-enabled option is being used by Percona Server. 
```

Author: updated.

> 

Reviewer 2: Please, I think it'd be a good idea to add a paragraph to introduce the topic. i.e. "Steps to make Percona Server use jemalloc with profiling enabled"

Author: Done.

    ./configure --enable-stats --enable-prof && make && make install
    MALLOC_CONF=prof:true 
    LD_PRELOAD=/usr/lib/libjemalloc.so

.. note:

    Ensure the ``libjemalloc.so`` exists in the LD path.

Usage
-------
To detect if jemalloc is set, run the following command:

.. sourcecode::  bash

    jemalloc_detected ON

To enable jemalloc profiling in a MySQL client, run the following command:

.. sourcecode:: bash

    set global jemalloc_profiling=on;

The :ref:`malloc_stats_totals` table returns the statistics, in bytes, of the memory usage. The command takes no parameters and returns the results as a table.

The following example commands display this result:

.. sourcecode:: mysql

    use performance_schema;

    SELECT * FROM malloc_stats_totals;
    +----+------------+------------+------------+-------------+------------+
    | id | ALLOCATION | MAPPED     | RESIDENT   | RETAINED    | METADATA   |
    +----+------------+------------+------------+-------------+------------+
    |  1 | 390977528  | 405291008  | 520167424  | 436813824   | 9933744    |
    +----+------------+------------+------------+-------------+------------+
    1 row in set (0.00 sec)

> The :ref:`malloc_stats` table returns the cumulative totals, in bytes, of allocation requests and allocation returns. The command takes no parameters and returns the results as a table.

Reviewer 2: ```suggestion
The :ref:`malloc_stats` table returns the cumulative totals, in bytes, of several statistics per type of arena. The command takes no parameters and returns the results as a table.
```

Author: updated.

The following example commands display this result:

.. sourcecode:: mysql

    use performance_schema;

    mysql> SELECT * FROM malloc_stats ORDER BY TYPE DESC LIMIT 3;
    +--------+-------------+-------------+-------------+-------------+
    | TYPE   | ALLOCATED   | NMALLOC     | NDALLOC     | NRESQUESTS  |
    +--------+-------------+-------------+-------------+-------------+
    | small  | 23578872    | 586156      | 0           | 2649417     |
    | large  | 367382528   | 2218        | 0           | 6355        |
    | huge   | 0           | 0           | 0           | |
    +--------+-------------+-------------+-------------+-------------+
    3 rows in set (0.00 sec)

> The profiling samples the ``malloc()`` calls and stores the sampled stack traces in a separate location in memory. These samples can be dumped into the filesystem. A dump returns a detailed view of the state of the memory. 

Reviewer 2: Maybe I would create a sub section for this i.e. `Dumping the memory profile`

Author: Added

As root, customize jemalloc with the following flags:

.. list-table::
    :widths: 20 40
    :header-rows: 1

    * - Option
      - Description
    * - `--enable-stats`
      - Enables statistics-gathering ability
    * - `--enable-prof`
      - Enables heap profiling and the ability to detect leaks. 

Using ``LD_PRELOAD``. Build the library, configure the malloc configuration with the ``prof:true`` string, and then use ``LD_PRELOAD`` to  preload the ``libjemalloc.so`` library. The libprocess ``MemoryProfiler`` class detects the library automatically and enables the profiling support.

The following is an example of the required commands:

.. sourcecode:: bash

    ./configure --enable-stats --enable-prof && make && make install
    MALLOC_CONF=prof:true 
    LD_PRELOAD=/usr/lib/libjemalloc.so

.. note:

    Ensure the ``libjemalloc.so`` exists in the LD path.

Use |Percona Server| with jemalloc with profiling enabled
----------------------------------------------------------

To detect if jemalloc is set, run the following command:

.. sourcecode::  bash

>     jemalloc_detected ON

Reviewer 3: I think this should be `SELECT @@jemalloc_detected;`?

Author: Fixed.
